### ISSUE TITLE
Consider a `-src`-style directive.

### ISSUE THREAD

**Issue by mikewest** [w3c/webappsec#184]
In https://lists.w3.org/Archives/Public/public-webappsec/2015Feb/0042.html, @pde suggested modifying the spec to allow whitelisting specific hosts rather than upgrading everything: "That way, if you have N third parties on a site, and (say) two of them provide images only, and don't support HTTPS at all, you can use the upgrade mechanism for scripts on the other N - 2 origins."

This would be pretty straightforward to do for a particular document, but would significantly complicate the inheritance mechanism. Pushing an enum down to a subframe is easy. Pushing a policy object is more complex.

**Comment by diracdeltas** [comment]
In https://github.com/w3ctag/spec-reviews/pull/54, we have been discussing whether this should be split into two whitelists: one for navigations, one for subresources.

I can't think of a use case where you would want these lists to be different. :/

Regardless, I think it would be good to have the ability to say, "Upgrade all (or some set of) navigations". Say I make a site that uses a third party redirector (like how twitter.com uses t.co). I would want navigations to be upgraded even though it's a third party.

**Comment by annevk** [comment]
Navigate and subresources are very different. E.g. I can ensure to only embed (potential) HTTPS resources, but I might have to provide pointers to non-HTTPS resources. Fairly typical scenario for WHATWG standards that have references to the W3C.

**Comment by diracdeltas** [comment]
@annevk I suppose that means in practice site operators will tend to be more conservative with navigation whitelist sets than with subresource whitelist sets. Ex: https://example.com is not sure whether example.org supports HTTPS on all resources, so it decides it's worth the risk to add 'example.org' to the subresource upgrade set but not the navigation upgrade set.

**Comment by annevk** [comment]
I was just replying to

> I can't think of a use case where you would want these lists to be different. :/

I think navigations are a mostly unexplored topic (with some exceptions, e.g. form-action) for CSP in general. Maybe once that is tackled this can also be addressed. But I'm not sure we need to do anything here right away.

**Comment by mikewest** [comment]
Thinking in the F2F about splitting `upgrade-insecure-navigations:` and `upgrade-insecure-subresource:` directives.

@hillbrad

**Comment by annevk** [comment]
Neither of those terms captures any kind of worker.

**Comment by mikewest** [comment]
Workers are subresources?

**Comment by annevk** [comment]
Nope, they are client requests. "subresource" is a confusing term to use anywhere actually. It's a request context. We should start using the terms outlined in this section: https://fetch.spec.whatwg.org/#image-request-context

**Comment by mikewest** [comment]
And I'm saying that the existence of `<link rel="subresource">` shouldn't mean that we can no longer use a commonly understood word to refer to a commonly understood thing (especially if there's no other commonly used way to refer to that thing).

I don't think there's going to be substantial confusion introduced if we distinguish between a "`subresource` request context" on the one hand, and a "subresource" on the other. *shrug*

**Comment by mikewest** [comment]
Based on the conversation at https://lists.w3.org/Archives/Public/public-webappsec/2015Aug/0042.html, I think we can punt this to The Future™.
